File: bindings/typescript/test/obol.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { mkdtempSync, copyFileSync, readFileSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import * as obol from "../src/index.ts";
// Imported from the public surface so the test also proves the export ships.
import { setPricingDir, clearPricingDir } from "../src/index.ts";

const HERE = dirname(fileURLToPath(import.meta.url));
const TESTDATA = join(HERE, "..", "..", "testdata"); // test -> typescript -> bindings, then /testdata
const OBOL_USAGE = join(TESTDATA, "obol-usage-mini.jsonl");
const ATIF_TRAJECTORY = join(TESTDATA, "atif-mini.json");

async function seed(): Promise<string> {
  const dir = mkdtempSync(join(tmpdir(), "obol-ts-"));
  copyFileSync(join(TESTDATA, "prices.json"), join(dir, "current.json"));
  await setPricingDir(dir);
  return dir;
}

test("version", async () => {
  // The binding reports the native lib's crate version; derive the
  // expectation from the workspace manifest so a release bump can't strand
  // a stale literal here (v0.7.0 broke CI exactly that way).
  const manifest = readFileSync(
    join(HERE, "..", "..", "..", "Cargo.toml"),
    "utf8",
  );
  const expected = /^version\s*=\s*"([^"]+)"/m.exec(manifest)?.[1];
  assert.ok(expected, "workspace Cargo.toml must declare a version");
  assert.equal(await obol.version(), expected);
});

test("estimatePath success", async () => {
  const dir = await seed();
  try {
    const est = await obol.estimatePath(OBOL_USAGE, "obol");
    assert.ok(est.total_usd > 0, `total_usd=${est.total_usd}`);
    assert.equal(est.pricing_as_of, "2026-06-05");
  } finally {
    rmSync(dir, { recursive: true, force: true });
    await clearPricingDir();
  }
});

test("estimatePath atif dialect prices the trajectory", async () => {
  const dir = await seed();
  try {
    const est = await obol.estimatePath(ATIF_TRAJECTORY, "atif");
    // opus by rates (36.75) + gpt-5.5 embedded cost (0.5) + unpriced model (0) = 37.25
    assert.ok(Math.abs(est.total_usd - 37.25) < 1e-9, `total_usd=${est.total_usd}`);
    // the unpriced model is surfaced, never a silent $0
    assert.ok(est.unpriced_models.includes("made-up-model-zzz"), JSON.stringify(est.unpriced_models));
  } finally {
    rmSync(dir, { recursive: true, force: true });
    await clearPricingDir();
  }
});

test("missing tables -> ObolError code 1", async () => {
  await setPricingDir("/nonexistent/obol-ts-xyz");
  try {
    await assert.rejects(
      () => obol.estimatePath(OBOL_USAGE, "obol"),
      (e: unknown) => e instanceof obol.ObolError && e.code === 1 && e.kind === "PricingTablesMissing",
    );
  } finally {
    await clearPricingDir();
  }
});

test("refresh rejects garbage as_of -> ObolError code 7", async () => {
  const dir = await seed();
  try {
    await assert.rejects(
      () => obol.refresh("Apr-2027"),
      (e: unknown) => e instanceof obol.ObolError && e.code === 7 && e.kind === "InvalidArgument",
    );
  } finally {
    rmSync(dir, { recursive: true, force: true });
    await clearPricingDir();
  }
});

test("unknown dialect -> ObolError code 7", async () => {
  const dir = await seed();
  try {
    await assert.rejects(
      () => obol.estimatePath(OBOL_USAGE, "banana" as obol.Dialect),
      (e: unknown) => e instanceof obol.ObolError && e.code === 7,
    );
  } finally {
    rmSync(dir, { recursive: true, force: true });
    await clearPricingDir();
  }
});

File: bindings/go/obol/obol_test.go
package obol

import (
	"os"
	"path/filepath"
	"testing"
)

func repoRoot(t *testing.T) string {
	wd, err := os.Getwd()
	if err != nil {
		t.Fatal(err)
	}
	return filepath.Join(wd, "..", "..", "..") // bindings/go/obol -> repo root
}

func testdata(t *testing.T) string { return filepath.Join(repoRoot(t), "bindings", "testdata") }

func seed(t *testing.T) {
	dir := t.TempDir()
	src, err := os.ReadFile(filepath.Join(testdata(t), "prices.json"))
	if err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(filepath.Join(dir, "current.json"), src, 0o644); err != nil {
		t.Fatal(err)
	}
	setPricingDir(t, dir)
}

func TestVersion(t *testing.T) {
	if Version() == "" {
		t.Fatal("empty version")
	}
}

func TestEstimatePath(t *testing.T) {
	seed(t)
	est, err := EstimatePath(filepath.Join(testdata(t), "claude-mini.jsonl"), "claude")
	if err != nil {
		t.Fatal(err)
	}
	if est.TotalUSD <= 0 {
		t.Fatalf("expected positive total, got %v", est.TotalUSD)
	}
	if est.PricingAsOf != "2026-06-05" {
		t.Fatalf("unexpected pricing_as_of %q", est.PricingAsOf)
	}
}

func TestMissingTablesIsError(t *testing.T) {
	setPricingDir(t, "/nonexistent/obol-go-xyz")
	_, err := EstimatePath(filepath.Join(testdata(t), "claude-mini.jsonl"), "claude")
	oe, ok := err.(*ObolError)
	if !ok || oe.Code != 1 {
		t.Fatalf("expected ObolError code 1, got %v", err)
	}
}

func TestRefreshRejectsGarbageAsOf(t *testing.T) {
	seed(t)
	_, err := Refresh("Apr-2027")
	oe, ok := err.(*ObolError)
	if !ok || oe.Code != 7 || oe.Kind != "InvalidArgument" {
		t.Fatalf("expected ObolError code 7 InvalidArgument, got %v", err)
	}
}

func TestUnknownDialectIsError(t *testing.T) {
	seed(t)
	_, err := EstimatePath(filepath.Join(testdata(t), "claude-mini.jsonl"), "banana")
	oe, ok := err.(*ObolError)
	if !ok || oe.Code != 7 {
		t.Fatalf("expected ObolError code 7, got %v", err)
	}
}
